src/audio/drc/drc_generic.c

```diff
@@ -465,255 +465,253 @@ static void drc_process_one_division(struct drc_state *state,
 	drc_compress_output(state, p, nbyte, nch);
 }
 
-#if CONFIG_FORMAT_S16LE
-static void drc_s16_default_pass(const struct comp_dev *dev,
-				 const struct audio_stream *source,
-				 struct audio_stream *sink,
-				 uint32_t frames)
+void drc_default_pass(const struct comp_dev *dev, const struct audio_stream *source,
+		      struct audio_stream *sink, uint32_t frames)
 {
-	int16_t *x;
-	int16_t *y;
-	int i;
-	int n = source->channels * frames;
+	audio_stream_copy(source, 0, sink, 0, frames * source->channels);
+}
 
-	for (i = 0; i < n; i++) {
-		x = audio_stream_read_frag_s16(source, i);
-		y = audio_stream_write_frag_s16(sink, i);
-		*y = *x;
-	}
+static inline void drc_pre_delay_index_inc(int *idx, int increment)
+{
+	*idx = (*idx + increment) & DRC_MAX_PRE_DELAY_FRAMES_MASK;
 }
-#endif /* CONFIG_FORMAT_S16LE */
 
-#if CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE
-static void drc_s32_default_pass(const struct comp_dev *dev,
-				 const struct audio_stream *source,
-				 struct audio_stream *sink,
-				 uint32_t frames)
+#if CONFIG_FORMAT_S16LE
+static void drc_delay_input_sample_s16(struct drc_state *state, const struct audio_stream *source,
+				       struct audio_stream *sink, int16_t **x, int16_t **y,
+				       int samples)
 {
-	int32_t *x;
-	int32_t *y;
+	int16_t *x1;
+	int16_t *y1;
+	int16_t *pd;
+	int pd_write_index, pd_read_index;
+	int nbuf, npcm, nfrm;
+	int ch;
 	int i;
-	int n = source->channels * frames;
+	int16_t *x0 = *x;
+	int16_t *y0 = *y;
+	int remaining_samples = samples;
+	int nch = source->channels;
 
-	for (i = 0; i < n; i++) {
-		x = audio_stream_read_frag_s32(source, i);
-		y = audio_stream_write_frag_s32(sink, i);
-		*y = *x;
+	while (remaining_samples) {
+		nbuf = DRC_BYTES_TO_S16_SAMPLES(audio_stream_bytes_without_wrap(source, x0));
+		npcm = MIN(remaining_samples, nbuf);
+		nbuf = DRC_BYTES_TO_S16_SAMPLES(audio_stream_bytes_without_wrap(sink, y0));
+		npcm = MIN(npcm, nbuf);
+		nfrm = npcm / nch;
```

Comment:
isn't this just audio_stream_frames?

Reply:
Yes it is but function audio_stream_frames_without_wrap() is slower than this way. It calls audio_stream_frame_bytes() -> get_frame_bytes() -> get_sample_bytes() that does a switch-case. There's division and multiplication. Audio stream frames function for a specific sample format would be better.

Reply:
Done, though when looking at this version code, this chunk looks quite generic and could go to audio_stream.h. What do you think?

```
		nbuf = audio_stream_samples_without_wrap_s24(source, x0);
		npcm = MIN(remaining_samples, nbuf);
		nbuf = audio_stream_samples_without_wrap_s24(sink, y0);
		npcm = MIN(npcm, nbuf);
		nfrm = npcm / nch;

```